packages/plugins/src/hardhat/simulate.ts

```diff
@@ -511,6 +455,129 @@ export const handleSimulationSuccess = async (
+          return initialResult
+        }
+
+        // A helper function that implements the timeout and retry logic for asynchronous calls to
+        // the Hardhat provider.
+        const invokeWithRetryAndSnapshot = async () => {
+          // We don't allow the 'hardhat_reset' RPC method to avoid an infinite loop bug caused by
+          // Hardhat. More context is in this pull request description:
+          // https://github.com/sphinx-labs/sphinx/pull/1565
+          if (args.length > 0 && args[0] === 'hardhat_reset') {
+            throw new Error(HardhatResetNotAllowedErrorMessage)
+          }
+
+          // Resolve the promise that we initially created.
+          let snapshotId = await snapshotIdPromise
+
+          for (
+            let attempt = 0;
+            attempt < simulationConstants.maxAttempts;
+            attempt++
+          ) {
+            try {
+              // If this is the first attempt, we use the `initialResult` instead of calling
+              // `target[prop](...args)` again to avoid unintended consequences of calling the
+              // method on the provider more than once.
+              const result =
+                attempt === 0 ? initialResult : target[prop](...args)
+
+              // Forward the call to the Hardhat provider. We include a timeout to ensure that an
+              // RPC provider with degraded service doesn't cause this call to hang indefinitely.
+              // See this pull request description for more info:
+              // https://github.com/sphinx-labs/sphinx/pull/1565
+              return await sphinxCoreUtils.callWithTimeout(
+                result,
+                simulationConstants.timeout,
+                getRpcRequestStalledErrorMessage(simulationConstants.timeout)
+              )
+            } catch (error) {
+              // The most likely reason that the call failed is a rate limit.
+
+              // NOTE: Don't include any RPC calls to the remote node in this 'catch' block because
+              // they may stall indefinitely if the user's RPC provider has degraded service. It's
+              // safe to call RPC methods that begin with 'evm_' or 'hardhat_' because these
+              // shouldn't be sent to the remote node, so there shouldn't be a risk of a rate limit
+              // occurring for these calls.
+
+              // We revert the Hardhat node state to ensure that there weren't any local state
+              // changes made by the failed RPC request. There should be no state changes because
+              // the call threw an error. This is a precaution against the "nonce too low" error,
+              // which is described in this pull request description:
+              // https://github.com/sphinx-labs/sphinx/pull/1565
+              const success = await target.send('evm_revert', [snapshotId])
+              if (!success) {
+                throw new InvariantError(`Failed to call 'evm_revert'.`)
+              }
```

Comment:
It's worth noting that this strategy could cause bugs if we're doing multiple state mutating requests simultaneously. Consider the following example:
1. Create a wallet with account A and submit a transaction with that account.
2. Create a wallet with account B and submit a second transaction without waiting for the previous transaction to get finalized. 

It's unclear what would happen if one of these transactions timed out. Depending on when the snapshot is taken for each transaction, you could potentially end up with a mixed-up state. 

This doesn't justify changing how this PR is implemented because we don't do anything like the above situation in the simulation, so this limitation won't cause us problems. However, I do think it's worth noting and documenting that this pattern prevents us from safely doing multiple calls that mutate state at the same time even if doing those calls simultaneously is perfectly valid.

Reply:
Yep good catch. Documented that. Also, will make a ticket for fixing this bug in Hardhat, since I think that'd resolve this issue as well as [this potential issue](https://github.com/sphinx-labs/sphinx/pull/1565#discussion_r1527251897) that I pointed out